### PR TITLE
1197: prepare beta-10.4 release

### LINKED ISSUE #1197: prepare beta-10.4 release
**What Needs to be Done?**
Prepare `beta-10.4` release.
Issues:
- #1195

### PR DESCRIPTION
**Pull request by cz4rs**
fixes #1197

### CODE REVIEW

File: tests/unit/runtime/test_initialization_with_args.cc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    // communicator is duplicated.
+    MPI_Comm comm = mpi_singleton->getComm();
+
+    static char prog_name[]{"vt_program"};
+    static char cli_argument[]{"--cli_argument=100"};
+    static char vt_no_terminate[]{"--vt_no_terminate"};
+    custom_args.emplace_back(&prog_name[0]);
+    custom_args.emplace_back(&cli_argument[0]);
+    custom_args.emplace_back(&vt_no_terminate[0]);
```
Comment (PhilMiller) [inline tests/unit/runtime/test_initialization_with_args.cc:72]: For strict standards compliance (i.e. if the parser in cli11 relies on it now or in the future), you should push a `nullptr` on the end of this vector.

https://stackoverflow.com/questions/16418932/is-argvargc-equal-to-null-pointer

**Comment by PhilMiller** [comment]
Tested successfully in EMPIRE with gcc. Nvidia and Intel are pending.

**Comment by PhilMiller** [comment]
Intel tests pass. Testing on CUDA is slow.

**Review by PhilMiller — APPROVED** [review]

**Comment by cz4rs** [comment]
@PhilMiller
Codacy and Travis jobs are stuck, other tests pass fine. As with the other PR, I cannot merge it myself (no permissions for override + I think merging into `1.0.0` is restricted as well)

if everything is ok, can you please merge this? (I will create the tag afterwards)

**Comment by cz4rs** [comment]
Regarding the extended tests, I have added the build definitions, but it seems like they have not been triggered.

Honestly I don't know if there should be much investigation put into this, since extended tests are failing regularly on `develop` anyways (I believe this was discussed at the last DARMA weekly, but there's no solution so far)
Screenshot from 2020-12-30 15-51-33

**Comment by PhilMiller** [comment]
We're looking at incorporating

- #915 Stacktrace generation
- A fix for missing deletion of https://github.com/DARMA-tasking/vt/pull/1113/files#diff-51dbee743b2faddab4d156c0fbf6707273c54f18e7b48a53480eefb455e0e2f0L100 in the backport onto beta 10.3
- #1107 assertions in CI builds

File: cmake-modules/CheckLinkerFlag.cmake [hunk 2]
```diff
@@ -0,0 +1,14 @@
+include(CheckCXXCompilerFlag)
+
+if(POLICY CMP0056)
+  # Honor link flags in try_compile() source-file signature.
+  cmake_policy(SET CMP0056 NEW)
+endif()
+
+# This is written because cmake does not support `check_linker_flag` until 3.18
```
Comment (PhilMiller) [inline cmake-modules/CheckLinkerFlag.cmake:8]: Does this need a version check around it? Was that concern brought up in #915?

**Comment by cz4rs** [comment]
note: #1098 has 2 PRs, but they were both done on `1.0.0-1098-enable-assertions-during-ci` branch

File: cmake-modules/CheckLinkerFlag.cmake (hunk 2 above)
Comment (cz4rs) [reply cmake-modules/CheckLinkerFlag.cmake:8]: there's a conversation in this PR regarding the implementation of this, but no mention of version check actually

Comment (PhilMiller) [reply cmake-modules/CheckLinkerFlag.cmake:8]: @lifflander Is this going to cause issues with very recent CMake versions?

Comment (cz4rs) [reply cmake-modules/CheckLinkerFlag.cmake:8]: I did some reading and it seems that overriding builtin functions is not recommended / potentially problematic (especially if someone would do the same in EMPIRE) - meaning we should only define it  `if(${CMAKE_VERSION} VERSION_LESS "3.18.0")` (?)

source:
https://crascit.com/2018/09/14/do-not-redefine-cmake-commands/

Comment (cz4rs) [reply cmake-modules/CheckLinkerFlag.cmake:8]: disclaimer: I don't feel like an expert on this, so a confirmation would be nice before opening a bug (I guess this should be fixed on develop as well)

Comment (cz4rs) [reply cmake-modules/CheckLinkerFlag.cmake:8]: we use `CMake` 3.18.4 in all `vt` containers:
```bash
ci/docker/ubuntu-18.04-clang-cpp.dockerfile:39:RUN ./cmake.sh 3.18.4
```

```bash
root@9d78a3e16284:/# cmake --version
cmake version 3.18.4
```

after removing / "ifdefing" our implementation of `check_linker_flag`, I am getting a following error:
```
root@9d78a3e16284:/# /vt/ci/build_cpp.sh /vt /build
(...)
CMake Error at cmake/set_linker_flags.cmake:3 (check_linker_flag):
  Unknown CMake command "check_linker_flag".
Call Stack (most recent call first):
  src/CMakeLists.txt:294 (include)

-- Configuring incomplete, errors occurred!
See also "/build/vt/CMakeFiles/CMakeOutput.log".
See also "/build/vt/CMakeFiles/CMakeError.log".
```

**Comment by PhilMiller** [comment]
This passes the full EMPIRE test suite in the `Debug` mode builds that unexpectedly failed. I'm re-running the test in other modes to not repeat my earlier mess.

**Comment by PhilMiller** [comment]
Manual and automatic testing of Release mode EMPIRE builds also passes. I'm re-running Intel 19 tests as well

**Comment by PhilMiller** [comment]
Intel 19 tests all pass as well. I think this is good to go.

**Comment by cz4rs** [comment]
> Intel 19 tests all pass as well. I think this is good to go.

:+1: 
@PhilMiller @lifflander 
If the CMake conversation above can be resolved as a non-issue, please merge this PR (I cannot do this myself - this is targeting release branch + Codacy is stuck again). I will create a tag & release notes afterwards.

**Comment by lifflander** [comment]
I think Travis CI needs to be removed for merging on release

File: cmake-modules/CheckLinkerFlag.cmake (hunk 2 above)
Comment (lifflander) [reply cmake-modules/CheckLinkerFlag.cmake:8]: I think this is fine because the policy is being invoked properly.

**Review by lifflander — APPROVED** [review]
Looks good to me. Let's merge.

**Pull request merged**